File: exercises/ddd_infrastructure_layer/java/src/main/java/domain/application/EmailAddress.java
package domain.application;
public class EmailAddress {
    private final String value;

    public EmailAddress(String value) {
        if (!isValidEmail(value)) {
            throw new IllegalArgumentException("Invalid email address.");
        }
        this.value = value;
    }

    public String getValue() {
        return value;
    }

    private boolean isValidEmail(String email) {
        // We would use more robust validation in a real application
        return email != null && email.contains("@") && email.contains(".");
    }
}

File: exercises/ddd_application_layer/typescript/src/domain/application/EmailAddress.ts
export class EmailAddress {
  private readonly _value: string;

  get value() { return this._value; }

  constructor(value: string) {
    if (!this.isValidEmail(value)) {
      throw new Error(`Invalid email address: ${value}`);
    }
    this._value = value;
  }

  private isValidEmail(value: string): boolean {
    // We would use more robust validation in a real application
    const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailPattern.test(value);
  }
}
